experiments/Compute.Experiments/Az.Compute.psm1

```diff
@@ -0,0 +1,331 @@
+#   * Implement new default formatter.
+#   * Lock down parameter list: should be good as is (if we add the option to set OS disk size).
+#   * Most of the parameters have "static" defaults: make these smart defaults, where applicable (e.g., location and open ports).
+#   * WinServer2016 is hardcoded: add parameters to allow people to use the full image provider, etc.
+#   * Import image providers, etc. from the "aliases.json".
+#   * Integrate into Nelson's build semantics.
+#   * Integrate app insights instrumentation so we can get good telemetry.
+
+#Requires -Modules AzureRM.Compute
+
+function New-AzVm {
+    param (
+        [Parameter(Mandatory = $true)] [string] $Name,
+        [Parameter(Mandatory = $true)] [string] $ResourceGroup,
+
+        # Generate a random as a hash of the name so it will be idempotent (tack on resource group?).
+        [Parameter(DontShow)]
+        $Random = $(Get-Hash $Name),
+
+        [string] $Location = "",
+
+        [string] $Image = "WinServer2016",
+        [string] $Size = "Standard_DS1_v2",
+
+        [string] $VnetName = "$($Name)Vnet",
+        [string] $SubnetAddressPrefix = "192.168.1.0/24",
+        [string] $VnetAddressPrefix = "192.168.0.0/16",
+
+        [string] $PublicIpName = "$($Name)PublicIp",
+        [string] $PublicIpDnsLabel = "$Name-$Random".ToLower(),
+        [string] $PublicIpAllocationMethod = "Static",
+        [int] $PublicIpIdleTimeoutInMinutes = 4,
+
+        [string] $NsgName = "$($Name)Nsg",
+        [int[]] $NsgOpenPorts = $null,
+
+        [string] $NicName = "$($Name)Nic"
+
+        # Storage - OS Disk Size.
+        # Compute: "this goes above and beyond the 80% scenario".
+    )
+
+    try {
+
+        # Build credentials.
+
+        $userName = $env:UserName;
+        $ptPassword = -join ((33..122) | Get-Random -Count 20 | ForEach-Object { [char]$_ });
+        $password = ConvertTo-SecureString $ptPassword -AsPlainText -Force;
+        $creds = New-Object System.Management.Automation.PSCredential ($env:UserName, $password);
+
+        # Set the smart defaults.
+
+        if($Location -eq "") {
+            # TODO: Infer the location somehow?
+            $Location = "westus2";
+        }
+
+        if($NsgOpenPorts -eq $null) {
+            # TODO: Infer the ports to open from the image types.
+            $NsgOpenPorts = @(3389,5985);
+        }
+
+        # Get image aliases.
+
+        # TODO: Properly set images from below.  Put this in its own "ensure" method?
+        $images = Invoke-WebRequest "https://raw.githubusercontent.com/Azure/azure-rest-api-specs/master/arm-compute/quickstart-templates/aliases.json" | ConvertFrom-Json;
+
+        # Set variables.
+
+        $fqdn = "$PublicIpDnsLabel.$Location.cloudapp.azure.com";
+
+        # Create!
+
+        Write-Info "Ensuring resource group...";
+        $rg = Use-ResourceGroup `
+            -ResourceGroup $ResourceGroup `
+            -Location $Location;
+
+        Write-Info "Ensuring Vnet...";
```

Comment:
For each of these, we should write relevant details about whether we are creating or using a resource, and the particulars (at least the rg + name)